### DISCUSSION TITLE
best coding practice to design a c code that should have a constructor in python bindings ?

### DISCUSSION THREAD

**Discussion by gogo2464** [swig/swig#2732] [Q&A] [answered]
Hello. I am coding a c alternative to pwntools to fix pwntools issues. I decided to provide bindings to be able to reuse my code in python.

Sadly the issue is that there is no constructor in C contrary to python. I would like to do

p = process('python')
p.recv(5)

in C. Sadly the C does not have any constructor nor class.

Should I use a struct with function pointer to create a pseudo class in C or is there a better programming practice for this case please?

**Comment by gogo2464** [comment]
I would not like to do:
```python
p.start(p)
p.recv(p, 5)
p.closeProcess(p)

# Destroy the instance
p.destroyProcess(p)
```

I would prefer to do:

```python
p.process(cmd)
p.recv(5)
p.closeProcess(p)

# Destroy the instance
p.destroyProcess()
```

instead

**Comment by gogo2464** [comment]
And could you provide an example please?

**Reply by ojwb** [reply, marked as answer]
You can use `%extend` to add constructors, destructors, methods to the wrapper for a C struct - there's an example in the manual:

https://www.swig.org/Doc4.1/SWIG.html#SWIG_adding_member_functions

**Reply by gogo2464** [reply]
Thank you very very much for your answer!!!!

In your opinion, could not it be even better to rewrite in C++ and to create C wrapper ?

https://github.com/swig/swig/issues/2733

**Reply by ojwb** [reply]
I'm not sure there's an absolute best approach as there are pros and cons to both approaches.

Though as you've seen in the linked issue, the C-as-a-SWIG-target-language is sadly still unmerged which typically makes it more awkward to deploy as you can't just install a package via whatever package management system you're using.

**Reply by gogo2464** [reply]
the dev of the branch specified before is my main concern. It is developped since a long time and still not merged. I also hope it does not lack documentation from https://github.com/swig/swig/pull/2086/files#diff-4d08eab8440ea799da34fe862f90b4dcbd22c9b320d1fd3f9a4aaefaa3f86c59 . 

For the sake of my project, I will choose C++with C bindings because my project already requires C++ dependencies.


Also my project will necessarely required C++ usage then I would like to directly code it in C++.

**Reply by gogo2464** [reply]
@ojwb thank you very much for your answer. I got an answer about C bindings from c++ betwee the time I asked and the time you answered. I have to test the bindings first.
